openedx_events/event_bus/avro/tests/test_avro.py

```diff
@@ -78,17 +152,86 @@ def test_all_events(self):
+        """
+        Test current version of events is forward compatible with stored schemas.
+
+        There's no assert because as long as the test doesn't raise an error, the new schema is
+        forward-compatible with the original.
+        """
+        for signal in OpenEdxPublicSignal.all_events():
+            if signal.event_type in KNOWN_UNSERIALIZABLE_SIGNALS:
+                continue
+            test_data = generate_test_data_for_signal(signal)
+            serializer = AvroSignalSerializer(signal)
+            schema_dict = serializer.schema
+
+            # write to bytes using current schema
+            current_out = io.BytesIO()
+            data_dict = serializer.to_dict(test_data)
+            schemaless_writer(current_out, schema_dict, data_dict)
+            current_out.seek(0)
+            current_event_bytes = current_out.read()
+
+            # get stored schema
+            schema_filename = f"{os.path.dirname(os.path.abspath(__file__))}/schemas/" \
+                              f"{signal.event_type.replace('.','+')}_schema.avsc"
+            try:
+                stored_schema = load_schema(schema_filename)
+            except SchemaRepositoryError:  # pragma: no cover
+                self.fail(f"Missing file {schema_filename}. If a new signal has been added, you may need to run the"
+                          f" generate_avro_schemas management command to save the signal schema.")
+
+            data_file_current = io.BytesIO(current_event_bytes)
+
+            # read bytes using stored schema
+            schemaless_reader(data_file_current, reader_schema=stored_schema, writer_schema=schema_dict)
+
+    def test_evolution_is_backward_compatible(self):
+        """
+        Test current version of events is backward compatible with stored schemas.
+
+        There's no assert because as long as the test doesn't raise an error, the new schema is
+        backward-compatible with the original.
+
+        Caveat: While we can go from an attrs class to a schema, we cannot go from a schema to an attrs class in the
+        same way because of custom serializers. Thus, when we are generating test data, we can only generate
+        the dictionary we would get after calling schema.to_dict.
+        """
+        for signal in OpenEdxPublicSignal.all_events():
+            if signal.event_type in KNOWN_UNSERIALIZABLE_SIGNALS:
+                continue
+            serializer = AvroSignalSerializer(signal)
+            schema_dict = serializer.schema
+
+            # get stored schema
+            schema_filename = f"{os.path.dirname(os.path.abspath(__file__))}/schemas/" \
+                              f"{signal.event_type.replace('.','+')}_schema.avsc"
+            try:
+                old_schema = load_schema(schema_filename)
+            except SchemaRepositoryError:  # pragma: no cover
+                self.fail(f"Missing file {schema_filename}. If a new signal has been added, you may need to run the"
```

Comment:
Thanks for this. We know this is an expected situation, and it just makes the experience much nicer for adding new events.